In [1]:
import pandas as pd

# Load the DataFrame using pandas
rce3_alone_clusters = pd.read_pickle("rce_pilot_3_alone_comp_cluster_ranges.pkl")

In [2]:
rce3_alone_clusters

video_name current_subject  \
0   20240320_171038_alone_comp_subj_4-2_and_4-3.1             4.2   
1   20240320_171038_alone_comp_subj_4-2_and_4-3.1             4.3   
2   20240322_120625_alone_comp_subj_3-3_and_3-4.1             3.3   
3   20240322_120625_alone_comp_subj_3-3_and_3-4.1             3.4   
4   20240322_160946_alone_comp_subj_4-3_and_4-4.1             4.3   
5   20240322_160946_alone_comp_subj_4-3_and_4-4.1             4.4   
6   20240323_122227_alone_comp_subj_5-2_and_5-3.1             5.2   
7   20240323_122227_alone_comp_subj_5-2_and_5-3.1             5.3   
8   20240323_144517_alone_comp_subj_3-1_and_3-4.1             3.1   
9   20240323_144517_alone_comp_subj_3-1_and_3-4.1             3.4   
10  20240323_165815_alone_comp_subj_4-2_and_4-4.1             4.2   
11  20240323_165815_alone_comp_subj_4-2_and_4-4.1             4.4   

                            cluster_index_ranges_dict  \
0   {7: [(0, 13), (587, 590), (434, 442), (449, 47...   
1   {7: [(0, 13), (587, 590), (434, 442), (449, 47...   
2   {3: [(66, 105), (242, 267), (412, 418), (174, ...   
3   {3: [(66, 105), (242, 267), (412, 418), (174, ...   
4   {7: [(0, 9), (13, 33), (91, 97), (389, 402), (...   
5   {7: [(0, 9), (13, 33), (91, 97), (389, 402), (...   
6   {7: [(0, 6), (236, 242), (438, 459), (545, 572...   
7   {7: [(0, 6), (236, 242), (438, 459), (545, 572...   
8   {5: [(0, 8), (25, 40), (55, 67), (133, 157), (...   
9   {5: [(0, 8), (25, 40), (55, 67), (133, 157), (...   
10  {5: [(0, 24), (40, 62), (112, 158), (186, 203)...   
11  {5: [(0, 24), (40, 62), (112, 158), (186, 203)...   

                            cluster_times_ranges_dict  \
0   {7: [(1600609, 1601302), (1630059, 1630198), (...   
1   {7: [(1600609, 1601302), (1630059, 1630198), (...   
2   {3: [(1527574, 1529514), (1536374, 1537622), (...   
3   {3: [(1527574, 1529514), (1536374, 1537622), (...   
4   {7: [(1438255, 1438740), (1438878, 1439918), (...   
5   {7: [(1438255, 1438740), (1438878, 1439918), (...   
6   {7: [(1600609, 1600887), (1612389, 1612667), (...   
7   {7: [(1600609, 1600887), (1612389, 1612667), (...   
8   {5: [(1683207, 1683623), (1684455, 1685217), (...   
9   {5: [(1683207, 1683623), (1684455, 1685217), (...   
10  {5: [(1698937, 1699907), (1700531, 1701431), (...   
11  {5: [(1698937, 1699907), (1700531, 1701431), (...   

                       cluster_timestamps_ranges_dict  \
0   {7: [(34079916, 34093775), (34668911, 34671683...   
1   {7: [(34079916, 34093775), (34668911, 34671683...   
2   {3: [(34169998, 34208802), (34346003, 34370949...   
3   {3: [(34169998, 34208802), (34346003, 34370949...   
4   {7: [(34096547, 34106248), (34109019, 34129808...   
5   {7: [(34096547, 34106248), (34109019, 34129808...   
6   {7: [(34070215, 34075759), (34305813, 34311357...   
7   {7: [(34070215, 34075759), (34305813, 34311357...   
8   {5: [(34957172, 34965487), (34982117, 34997362...   
9   {5: [(34957172, 34965487), (34982117, 34997362...   
10  {5: [(36182281, 36201683), (36214156, 36232172...   
11  {5: [(36182281, 36201683), (36214156, 36232172...   

                 trial_cluster_timestamps_ranges_dict  \
0   {'lose_7': [(34079916, 34093775), (34668911, 3...   
1   {'win_7': [(34079916, 34093775), (34668911, 34...   
2   {'tie_3': [(34169998, 34208802), (34346003, 34...   
3   {'tie_3': [(34169998, 34208802), (34346003, 34...   
4   {'tie_7': [(34096547, 34106248), (34109019, 34...   
5   {'tie_7': [(34096547, 34106248), (34109019, 34...   
6   {'lose_7': [(34070215, 34075759), (34305813, 3...   
7   {'win_7': [(34070215, 34075759), (34305813, 34...   
8   {'lose_5': [(34957172, 34965487), (34982117, 3...   
9   {'win_5': [(34957172, 34965487), (34982117, 34...   
10  {'win_5': [(36182281, 36201683), (36214156, 36...   
11  {'lose_5': [(36182281, 36201683), (36214156, 3...   

                      trial_cluster_times_ranges_dict  \
0   {'lose_7': [(1600609, 1601302), (1630059, 1630...   
1   {'win_7': [(1600609, 1601302), (1630059, 16301...   
2 

In [15]:
# Access each dictionary cell in row 0
cluster_index_ranges = rce3_alone_clusters.at[0, 'cluster_index_ranges_dict']
cluster_times_ranges = rce3_alone_clusters.at[0, 'cluster_times_ranges_dict']
cluster_timestamps_ranges = rce3_alone_clusters.at[0, 'cluster_timestamps_ranges_dict']
trial_cluster_timestamps_ranges = rce3_alone_clusters.at[0, 'trial_cluster_timestamps_ranges_dict']
trial_cluster_times_ranges = rce3_alone_clusters.at[0, 'trial_cluster_times_ranges_dict']

type(cluster_index_ranges)

dict

In [7]:
cluster_index_ranges.keys()

dict_keys([np.int64(7), np.int64(1), np.int64(5), np.int64(2), np.int64(3), np.int64(6), np.int64(0)])

In [17]:
cluster_index_ranges.keys()

dict_keys([np.int64(7), np.int64(1), np.int64(5), np.int64(2), np.int64(3), np.int64(6), np.int64(0)])

In [18]:
cluster_times_ranges.keys()

dict_keys([np.int64(7), np.int64(1), np.int64(5), np.int64(2), np.int64(3), np.int64(6), np.int64(0)])

In [19]:
cluster_timestamps_ranges.keys()

dict_keys([np.int64(7), np.int64(1), np.int64(5), np.int64(2), np.int64(3), np.int64(6), np.int64(0)])

In [20]:
trial_cluster_timestamps_ranges.keys()

dict_keys(['lose_7', 'lose_1', 'lose_5', 'lose_2', 'lose_3', 'lose_6', 'lose_0', 'tie_5', 'tie_1', 'tie_7', 'tie_3', 'tie_2', 'tie_0'])

In [21]:
trial_cluster_times_ranges.keys()

dict_keys(['lose_7', 'lose_1', 'lose_5', 'lose_2', 'lose_3', 'lose_6', 'lose_0', 'tie_5', 'tie_1', 'tie_7', 'tie_3', 'tie_2', 'tie_0'])

In [22]:
import pandas as pd
from collections import Counter

# Assuming rce3_alone_clusters is your DataFrame
cluster_index_ranges_all = rce3_alone_clusters['cluster_index_ranges_dict'].tolist()

# Flatten the list of dictionaries and extract keys
all_keys = [key for cluster_dict in cluster_index_ranges_all for key in cluster_dict.keys()]

# Count the occurrences of each unique key
key_counts = Counter(all_keys)

# Convert to a dictionary with string keys (for consistency)
key_counts_str = {str(key): count for key, count in key_counts.items()}

print(key_counts_str)  # Output the results in the desired format

{'7': 12, '1': 12, '5': 12, '2': 12, '3': 12, '6': 12, '0': 12, '4': 2}


In [23]:
def sort_cluster_index_ranges(cluster_index_ranges_dict):
    """Sorts the ranges within a cluster_index_ranges_dict.

    Args:
        cluster_index_ranges_dict: A dictionary where keys are cluster IDs and values
            are lists of (start, end) tuples representing ranges.

    Returns:
        A new dictionary with the same keys but sorted ranges.
    """

    sorted_ranges_dict = {}
    for cluster_id, ranges in cluster_index_ranges_dict.items():
        sorted_ranges_dict[cluster_id] = sorted(ranges)
    return sorted_ranges_dict

# Example Usage (assuming you have the corrected 'cluster_index_ranges' variable)
sorted_ranges_dict = sort_cluster_index_ranges(cluster_index_ranges)

# To get the sorted ranges for cluster ID 7:
sorted_ranges_for_7 = sorted_ranges_dict[7]
print(sorted_ranges_for_7) 


[(0, 10), (0, 13), (0, 16), (2, 21), (5, 20), (10, 13), (10, 14), (12, 13), (20, 24), (24, 25), (27, 28), (30, 51), (32, 60), (36, 54), (45, 49), (47, 54), (57, 67), (67, 73), (70, 91), (71, 77), (71, 85), (90, 101), (92, 95), (92, 104), (100, 136), (102, 115), (107, 111), (119, 138), (121, 134), (132, 135), (132, 157), (134, 135), (135, 158), (147, 157), (148, 173), (149, 152), (158, 168), (160, 163), (162, 167), (167, 169), (169, 176), (170, 180), (173, 177), (180, 181), (186, 203), (194, 197), (197, 201), (198, 212), (200, 203), (215, 220), (219, 228), (224, 228), (224, 233), (230, 232), (232, 233), (243, 260), (255, 270), (263, 265), (264, 267), (280, 289), (285, 289), (286, 290), (310, 328), (312, 315), (331, 336), (341, 348), (342, 344), (346, 349), (349, 367), (355, 363), (355, 372), (357, 361), (365, 370), (365, 374), (374, 392), (376, 386), (379, 389), (384, 389), (385, 391), (387, 405), (389, 395), (395, 407), (399, 419), (403, 406), (417, 425), (419, 451), (425, 462), (425, 

First timepoint in Excel sheet: 34,280,576  
First timepoint in cluster_index_ranges: 0  
First timepoint in cluster_times_ranges: 1,600,609  
First timepoint in cluster_timestamps_ranges: 34,079,916  
First timepoint in trial_cluster_timestamps_ranges: 34,079,916  
First timepoint in trial_cluster_times_ranges: 1,600,609  

<br>
<br>

Last timepoint in Excel sheet: 61,480,909  
Last timepoint in cluster_index_ranges: 591  
Last timepoint in cluster_times_ranges: 2,978,094  
Last timepoint in cluster_timestamps_ranges: 61,108,578  
Last timepoint in trial_cluster_timestamps_ranges: 61,108,578  
Last timepoint in trial_cluster_times_ranges: 2,952,043

In [39]:
61108578 // 20

3055428

First win timepoint (Excel // 20): 1,713,814  
First timepoint cluster_timestamps_ranges // 20: 1,703,995

Last win timepoint (Excel // 20): 2,788,828  
Last timepoint cluster_timestamps_ranges // 20: 3,055,428

In [40]:
61108578 / 20

3055428.9

In [41]:
round(61108578 / 20)

3055429

In [42]:
import pandas as pd
import numpy as np
import ast
import pickle
from pathlib import Path
import matplotlib.pyplot as plt
import multirecording_spikeanalysis as spike

In [69]:
df = rce3_alone_clusters

In [ ]:
# Initialize an empty list to collect data for the new DataFrame
new_df_data = []

for _, row in df.iterrows():
    subject = row['current_subject'].replace('.', '-')
    session_dir = row['session_dir']
    base_session_dir = session_dir.split('_subj_')[0]
    
    timestamps = row['trial_cluster_timestamps_ranges']

    for key, ranges in timestamps.items():  # Iterate over the dictionary, not individual keys
        new_df.append({
            'cluster': int(key),
            'timestamps': [[int(x) for x in range] for range in ranges]  # Iterate over each range tuple
        })
    
# Convert list to DataFrame
new_df = pd.DataFrame(new_df_data)
new_df = new_df.drop_duplicates()

In [ ]:
# Initialize an empty list to collect data for the new DataFrame
new_df_data = []

for _, row in df3.iterrows():
    session_dir = row['session_dir']
    subjects = row['all_subjects']
    condition = row['condition ']

    # Split session_dir on '_subj_' and take the first part only
    # This ensures everything after '_subj_' is ignored
    base_session_dir = session_dir.split('_subj_')[0]

    for subject in subjects:
        subject_formatted = subject.replace('.', '-')
        # Append formatted subject to the base session_dir correctly
        subj_recording = f"{base_session_dir}_subj_{subject_formatted}"
        new_df_data.append({
            'session_dir': session_dir,
            'subject': subject,
            'subj_recording': subj_recording,
            'condition': condition if condition in ['rewarded', 'omission', 'both_rewarded', 'tie'] else ('win' if str(condition) == str(subject) else 'lose'),
            'tone_start_timestamp': row['tone_start_timestamp'],
            'tone_stop_timestamp': row['tone_stop_timestamp']
        })

# Convert list to DataFrame
new_df = pd.DataFrame(new_df_data)
new_df = new_df.drop_duplicates()

In [66]:
new_df = []

for key, ranges in cluster_timestamps_ranges.items():  # Iterate over the dictionary, not individual keys
    new_df.append({
        'cluster': int(key),
        'timestamps': [[int(x) for x in range] for range in ranges]  # Iterate over each range tuple
    })

In [57]:
abc = 123
type(abc)

int

In [58]:
abc2 = np.int32(123)
type(abc2)

numpy.int32

In [60]:
abc3 = int(abc2)
type(abc3)

int

In [73]:
def extract_cluster_data(df, cluster_type="cluster_timestamps_ranges_dict"):
    """Extracts cluster data from a DataFrame.

    Args:
        df: The input DataFrame.
        cluster_type: The column name containing the cluster data. 
                      Defaults to 'trial_cluster_timestamps_ranges_dict'.

    Returns:
        A new DataFrame with columns: 'recording', 'subject', 'cluster', and 'ranges'.
    """
    new_df_data = []

    for _, row in df.iterrows():
        base_session_dir = row['session_dir'].split('_subj_')[0]
        subject = row['current_subject'].replace('.', '-')
        recording = f"{base_session_dir}_subj_{subject}"
        
        # Fetch and ensure cluster data is a dictionary
        cluster_data = row[cluster_type]
        if not isinstance(cluster_data, dict):
            print(f"Warning: Invalid cluster data for {recording}: {cluster_data}")
            continue  # Skip this row if data is invalid

        for cluster_id, ranges in cluster_data.items():
            new_df_data.append({
                'recording': recording,
                'subject': subject,
                'cluster': int(cluster_id),
                'ranges': ranges,  # Store the ranges as a list of tuples
            })

    return pd.DataFrame(new_df_data)
    


# Example usage
new_df = extract_cluster_data(df)

# Optionally, sort by recording and cluster to organize the data
new_df = new_df.sort_values(by=['recording', 'cluster'])

new_df

recording subject  cluster  \
6   20240320_171038_alone_comp_subj_4-2     4-2        0   
1   20240320_171038_alone_comp_subj_4-2     4-2        1   
3   20240320_171038_alone_comp_subj_4-2     4-2        2   
4   20240320_171038_alone_comp_subj_4-2     4-2        3   
2   20240320_171038_alone_comp_subj_4-2     4-2        5   
..                                  ...     ...      ...   
81  20240323_165815_alone_comp_subj_4-4     4-4        2   
82  20240323_165815_alone_comp_subj_4-4     4-4        3   
79  20240323_165815_alone_comp_subj_4-4     4-4        5   
80  20240323_165815_alone_comp_subj_4-4     4-4        6   
83  20240323_165815_alone_comp_subj_4-4     4-4        7   

                                               ranges  
6   [(43031248, 43035406), (43054808, 43058965), (...  
1   [(34095161, 34097933), (34142281, 34182471), (...  
3   [(34138123, 34140895), (34186628, 34204645), (...  
4   [(34272552, 34276710), (34663367, 34667525), (...  
2   [(34099319, 34138123), (34183857, 34185242), (...  
..                                                ...  
81  [(36233558, 36276520), (36325025, 36349971), (...  
82  [(36380460, 36392933), (36463612, 36498259), (...  
79  [(36182281, 36201683), (36214156, 36232172), (...  
80  [(36203069, 36214156), (37164862, 37180107), (...  
83  [(36501031, 36507960), (36603585, 36617444), (...  

[86 rows x 4 columns]

In [71]:
df

video_name current_subject  \
0   20240320_171038_alone_comp_subj_4-2_and_4-3.1             4.2   
1   20240320_171038_alone_comp_subj_4-2_and_4-3.1             4.3   
2   20240322_120625_alone_comp_subj_3-3_and_3-4.1             3.3   
3   20240322_120625_alone_comp_subj_3-3_and_3-4.1             3.4   
4   20240322_160946_alone_comp_subj_4-3_and_4-4.1             4.3   
5   20240322_160946_alone_comp_subj_4-3_and_4-4.1             4.4   
6   20240323_122227_alone_comp_subj_5-2_and_5-3.1             5.2   
7   20240323_122227_alone_comp_subj_5-2_and_5-3.1             5.3   
8   20240323_144517_alone_comp_subj_3-1_and_3-4.1             3.1   
9   20240323_144517_alone_comp_subj_3-1_and_3-4.1             3.4   
10  20240323_165815_alone_comp_subj_4-2_and_4-4.1             4.2   
11  20240323_165815_alone_comp_subj_4-2_and_4-4.1             4.4   

                            cluster_index_ranges_dict  \
0   {7: [(0, 13), (587, 590), (434, 442), (449, 47...   
1   {7: [(0, 13), (587, 590), (434, 442), (449, 47...   
2   {3: [(66, 105), (242, 267), (412, 418), (174, ...   
3   {3: [(66, 105), (242, 267), (412, 418), (174, ...   
4   {7: [(0, 9), (13, 33), (91, 97), (389, 402), (...   
5   {7: [(0, 9), (13, 33), (91, 97), (389, 402), (...   
6   {7: [(0, 6), (236, 242), (438, 459), (545, 572...   
7   {7: [(0, 6), (236, 242), (438, 459), (545, 572...   
8   {5: [(0, 8), (25, 40), (55, 67), (133, 157), (...   
9   {5: [(0, 8), (25, 40), (55, 67), (133, 157), (...   
10  {5: [(0, 24), (40, 62), (112, 158), (186, 203)...   
11  {5: [(0, 24), (40, 62), (112, 158), (186, 203)...   

                            cluster_times_ranges_dict  \
0   {7: [(1600609, 1601302), (1630059, 1630198), (...   
1   {7: [(1600609, 1601302), (1630059, 1630198), (...   
2   {3: [(1527574, 1529514), (1536374, 1537622), (...   
3   {3: [(1527574, 1529514), (1536374, 1537622), (...   
4   {7: [(1438255, 1438740), (1438878, 1439918), (...   
5   {7: [(1438255, 1438740), (1438878, 1439918), (...   
6   {7: [(1600609, 1600887), (1612389, 1612667), (...   
7   {7: [(1600609, 1600887), (1612389, 1612667), (...   
8   {5: [(1683207, 1683623), (1684455, 1685217), (...   
9   {5: [(1683207, 1683623), (1684455, 1685217), (...   
10  {5: [(1698937, 1699907), (1700531, 1701431), (...   
11  {5: [(1698937, 1699907), (1700531, 1701431), (...   

                       cluster_timestamps_ranges_dict  \
0   {7: [(34079916, 34093775), (34668911, 34671683...   
1   {7: [(34079916, 34093775), (34668911, 34671683...   
2   {3: [(34169998, 34208802), (34346003, 34370949...   
3   {3: [(34169998, 34208802), (34346003, 34370949...   
4   {7: [(34096547, 34106248), (34109019, 34129808...   
5   {7: [(34096547, 34106248), (34109019, 34129808...   
6   {7: [(34070215, 34075759), (34305813, 34311357...   
7   {7: [(34070215, 34075759), (34305813, 34311357...   
8   {5: [(34957172, 34965487), (34982117, 34997362...   
9   {5: [(34957172, 34965487), (34982117, 34997362...   
10  {5: [(36182281, 36201683), (36214156, 36232172...   
11  {5: [(36182281, 36201683), (36214156, 36232172...   

                 trial_cluster_timestamps_ranges_dict  \
0   {'lose_7': [(34079916, 34093775), (34668911, 3...   
1   {'win_7': [(34079916, 34093775), (34668911, 34...   
2   {'tie_3': [(34169998, 34208802), (34346003, 34...   
3   {'tie_3': [(34169998, 34208802), (34346003, 34...   
4   {'tie_7': [(34096547, 34106248), (34109019, 34...   
5   {'tie_7': [(34096547, 34106248), (34109019, 34...   
6   {'lose_7': [(34070215, 34075759), (34305813, 3...   
7   {'win_7': [(34070215, 34075759), (34305813, 34...   
8   {'lose_5': [(34957172, 34965487), (34982117, 3...   
9   {'win_5': [(34957172, 34965487), (34982117, 34...   
10  {'win_5': [(36182281, 36201683), (36214156, 36...   
11  {'lose_5': [(36182281, 36201683), (36214156, 3...   

                      trial_cluster_times_ranges_dict  \
0   {'lose_7': [(1600609, 1601302), (1630059, 1630...   
1   {'win_7': [(1600609, 1601302), (1630059, 16301...   
2 

In [74]:
array1 = new_df.at[0, 'ranges']

In [80]:
array2 = [
    [int(start // 20), int(end // 20)]  # Divide both start and end timestamps
    for start, end in array1 
]

In [81]:
array2

[[1703995, 1704688],
 [1733445, 1733584],
 [1775783, 1776199],
 [1776546, 1778070],
 [1817542, 1817914],
 [1818399, 1819230],
 [1820616, 1820686],
 [1821864, 1822556],
 [1825675, 1825883],
 [1828585, 1828862],
 [1829070, 1830040],
 [1831495, 1831842],
 [1833921, 1834752],
 [1942711, 1942919],
 [2042286, 2042355],
 [2136387, 2136733],
 [2140804, 2141930],
 [2151354, 2151493],
 [2151839, 2152671],
 [2153017, 2153502],
 [2155304, 2158006],
 [2163550, 2163619],
 [2201938, 2202077],
 [2211709, 2211917],
 [2234506, 2234645],
 [2242128, 2242406],
 [2321400, 2322578],
 [2369074, 2369213],
 [2376142, 2376489],
 [2376835, 2377251],
 [2378359, 2378637],
 [2379953, 2380300],
 [2384873, 2384942],
 [2387783, 2388338],
 [2473638, 2473777],
 [2474331, 2474539],
 [2474955, 2475925],
 [2476479, 2476618],
 [2478835, 2479099],
 [2479768, 2480013],
 [2480221, 2480429],
 [2482993, 2483478],
 [2486457, 2487358],
 [2487774, 2488605],
 [2489160, 2489368],
 [2490268, 2492139],
 [2493941, 2494287],
 [2494773, 24

In [83]:
df

video_name current_subject  \
0   20240320_171038_alone_comp_subj_4-2_and_4-3.1             4.2   
1   20240320_171038_alone_comp_subj_4-2_and_4-3.1             4.3   
2   20240322_120625_alone_comp_subj_3-3_and_3-4.1             3.3   
3   20240322_120625_alone_comp_subj_3-3_and_3-4.1             3.4   
4   20240322_160946_alone_comp_subj_4-3_and_4-4.1             4.3   
5   20240322_160946_alone_comp_subj_4-3_and_4-4.1             4.4   
6   20240323_122227_alone_comp_subj_5-2_and_5-3.1             5.2   
7   20240323_122227_alone_comp_subj_5-2_and_5-3.1             5.3   
8   20240323_144517_alone_comp_subj_3-1_and_3-4.1             3.1   
9   20240323_144517_alone_comp_subj_3-1_and_3-4.1             3.4   
10  20240323_165815_alone_comp_subj_4-2_and_4-4.1             4.2   
11  20240323_165815_alone_comp_subj_4-2_and_4-4.1             4.4   

                            cluster_index_ranges_dict  \
0   {7: [(0, 13), (587, 590), (434, 442), (449, 47...   
1   {7: [(0, 13), (587, 590), (434, 442), (449, 47...   
2   {3: [(66, 105), (242, 267), (412, 418), (174, ...   
3   {3: [(66, 105), (242, 267), (412, 418), (174, ...   
4   {7: [(0, 9), (13, 33), (91, 97), (389, 402), (...   
5   {7: [(0, 9), (13, 33), (91, 97), (389, 402), (...   
6   {7: [(0, 6), (236, 242), (438, 459), (545, 572...   
7   {7: [(0, 6), (236, 242), (438, 459), (545, 572...   
8   {5: [(0, 8), (25, 40), (55, 67), (133, 157), (...   
9   {5: [(0, 8), (25, 40), (55, 67), (133, 157), (...   
10  {5: [(0, 24), (40, 62), (112, 158), (186, 203)...   
11  {5: [(0, 24), (40, 62), (112, 158), (186, 203)...   

                            cluster_times_ranges_dict  \
0   {7: [(1600609, 1601302), (1630059, 1630198), (...   
1   {7: [(1600609, 1601302), (1630059, 1630198), (...   
2   {3: [(1527574, 1529514), (1536374, 1537622), (...   
3   {3: [(1527574, 1529514), (1536374, 1537622), (...   
4   {7: [(1438255, 1438740), (1438878, 1439918), (...   
5   {7: [(1438255, 1438740), (1438878, 1439918), (...   
6   {7: [(1600609, 1600887), (1612389, 1612667), (...   
7   {7: [(1600609, 1600887), (1612389, 1612667), (...   
8   {5: [(1683207, 1683623), (1684455, 1685217), (...   
9   {5: [(1683207, 1683623), (1684455, 1685217), (...   
10  {5: [(1698937, 1699907), (1700531, 1701431), (...   
11  {5: [(1698937, 1699907), (1700531, 1701431), (...   

                       cluster_timestamps_ranges_dict  \
0   {7: [(34079916, 34093775), (34668911, 34671683...   
1   {7: [(34079916, 34093775), (34668911, 34671683...   
2   {3: [(34169998, 34208802), (34346003, 34370949...   
3   {3: [(34169998, 34208802), (34346003, 34370949...   
4   {7: [(34096547, 34106248), (34109019, 34129808...   
5   {7: [(34096547, 34106248), (34109019, 34129808...   
6   {7: [(34070215, 34075759), (34305813, 34311357...   
7   {7: [(34070215, 34075759), (34305813, 34311357...   
8   {5: [(34957172, 34965487), (34982117, 34997362...   
9   {5: [(34957172, 34965487), (34982117, 34997362...   
10  {5: [(36182281, 36201683), (36214156, 36232172...   
11  {5: [(36182281, 36201683), (36214156, 36232172...   

                 trial_cluster_timestamps_ranges_dict  \
0   {'lose_7': [(34079916, 34093775), (34668911, 3...   
1   {'win_7': [(34079916, 34093775), (34668911, 34...   
2   {'tie_3': [(34169998, 34208802), (34346003, 34...   
3   {'tie_3': [(34169998, 34208802), (34346003, 34...   
4   {'tie_7': [(34096547, 34106248), (34109019, 34...   
5   {'tie_7': [(34096547, 34106248), (34109019, 34...   
6   {'lose_7': [(34070215, 34075759), (34305813, 3...   
7   {'win_7': [(34070215, 34075759), (34305813, 34...   
8   {'lose_5': [(34957172, 34965487), (34982117, 3...   
9   {'win_5': [(34957172, 34965487), (34982117, 34...   
10  {'win_5': [(36182281, 36201683), (36214156, 36...   
11  {'lose_5': [(36182281, 36201683), (36214156, 3...   

                      trial_cluster_times_ranges_dict  \
0   {'lose_7': [(1600609, 1601302), (1630059, 1630...   
1   {'win_7': [(1600609, 1601302), (1630059, 16301...   
2 

In [84]:
new_df

recording subject  cluster  \
6   20240320_171038_alone_comp_subj_4-2     4-2        0   
1   20240320_171038_alone_comp_subj_4-2     4-2        1   
3   20240320_171038_alone_comp_subj_4-2     4-2        2   
4   20240320_171038_alone_comp_subj_4-2     4-2        3   
2   20240320_171038_alone_comp_subj_4-2     4-2        5   
..                                  ...     ...      ...   
81  20240323_165815_alone_comp_subj_4-4     4-4        2   
82  20240323_165815_alone_comp_subj_4-4     4-4        3   
79  20240323_165815_alone_comp_subj_4-4     4-4        5   
80  20240323_165815_alone_comp_subj_4-4     4-4        6   
83  20240323_165815_alone_comp_subj_4-4     4-4        7   

                                               ranges  
6   [(43031248, 43035406), (43054808, 43058965), (...  
1   [(34095161, 34097933), (34142281, 34182471), (...  
3   [(34138123, 34140895), (34186628, 34204645), (...  
4   [(34272552, 34276710), (34663367, 34667525), (...  
2   [(34099319, 34138123), (34183857, 34185242), (...  
..                                                ...  
81  [(36233558, 36276520), (36325025, 36349971), (...  
82  [(36380460, 36392933), (36463612, 36498259), (...  
79  [(36182281, 36201683), (36214156, 36232172), (...  
80  [(36203069, 36214156), (37164862, 37180107), (...  
83  [(36501031, 36507960), (36603585, 36617444), (...  

[86 rows x 4 columns]

In [86]:
import pandas as pd
import numpy as np

def extract_cluster_data(df, cluster_type="cluster_timestamps_ranges_dict"):
    """Extracts cluster data from a DataFrame and reformats it into a structured DataFrame.

    Args:
        df: The input DataFrame.
        cluster_type: The column name containing the cluster data. Defaults to 'cluster_timestamps_ranges_dict'.

    Returns:
        A new DataFrame with columns: 'recording', 'subject', 'cluster', and 'ranges'.
    """
    new_df_data = []

    for _, row in df.iterrows():
        # Construct the base recording string from session_dir and subject
        base_session_dir = row['session_dir'].split('_subj_')[0]
        subject_formatted = row['current_subject'].replace('.', '-')
        recording = f"{base_session_dir}_subj_{subject_formatted}"

        # Fetch and validate cluster data as a dictionary
        cluster_data = row[cluster_type]
        if not isinstance(cluster_data, dict):
            print(f"Warning: Invalid cluster data for {recording}: {cluster_data}")
            continue

        # Append data for each cluster to the new DataFrame list
        for cluster_id, ranges in cluster_data.items():
            new_df_data.append({
                'recording': recording,
                'subject': subject_formatted,
                'cluster': int(cluster_id),
                'ranges': np.array(ranges, dtype=np.int64)  # Convert ranges to numpy arrays for efficient processing
            })

    return pd.DataFrame(new_df_data)

# Usage example (assuming df is already defined)
new_df = extract_cluster_data(df)

new_df

recording subject  cluster  \
0   20240320_171038_alone_comp_subj_4-2     4-2        7   
1   20240320_171038_alone_comp_subj_4-2     4-2        1   
2   20240320_171038_alone_comp_subj_4-2     4-2        5   
3   20240320_171038_alone_comp_subj_4-2     4-2        2   
4   20240320_171038_alone_comp_subj_4-2     4-2        3   
..                                  ...     ...      ...   
81  20240323_165815_alone_comp_subj_4-4     4-4        2   
82  20240323_165815_alone_comp_subj_4-4     4-4        3   
83  20240323_165815_alone_comp_subj_4-4     4-4        7   
84  20240323_165815_alone_comp_subj_4-4     4-4        1   
85  20240323_165815_alone_comp_subj_4-4     4-4        0   

                                               ranges  
0   [[34079916, 34093775], [34668911, 34671683], [...  
1   [[34095161, 34097933], [34142281, 34182471], [...  
2   [[34099319, 34138123], [34183857, 34185242], [...  
3   [[34138123, 34140895], [34186628, 34204645], [...  
4   [[34272552, 34276710], [34663367, 34667525], [...  
..                                                ...  
81  [[36233558, 36276520], [36325025, 36349971], [...  
82  [[36380460, 36392933], [36463612, 36498259], [...  
83  [[36501031, 36507960], [36603585, 36617444], [...  
84  [[37249400, 37257716], [37293748, 37311764], [...  
85  [[40691900, 40698830], [44017987, 44077580], [...  

[86 rows x 4 columns]